refactor: :recycle: update `extract_resource_properties()` to take `pl.DataFrame`

signekb opened the pull request:
## Description

This PR updates `extract_resource_properties()` to take a `pl.DataFrame` instead of a `Path` to raw data. 

In doing this, I:

- added a mapping from `POLARS_TO_FRICTIONLESS` (which I’m not entirely sure about, so please take an in-depth look at that ☀️ ). 
- updated the resources guide. 
- updated the all types example to have a `my_list` field and updated the `dtypes` of `my_none` to `pl.Null` and `my_any` to `pl.Unknown`. 
- removed `frictionless py` as a dependency 🎉 

Closes #1203 

<!-- Select quick/in-depth as necessary -->
This PR needs an in-depth review.

## Checklist

- [X] Added or updated tests
- [X] Updated documentation
- [X] Ran `just run-all`

signekb committed:
refactor: :recycle: update `extract_resource_properties()` to take `pl.DataFrame`

Instead of a path to raw data.

signekb committed:
test: :white_check_mark: update tests to match updates to function

signekb committed:
refactor: :recycle: update `all_types` examples to include list and update dtypes

signekb committed:
docs: :memo: update to extract from df; remove mentions of "raw"; use print for nested objects

signekb committed:
docs: :art: format `qmd`

signekb committed:
build: :heavy_minus_sign: remove `frictionless` from dependencies

pre-commit-ci[bot] committed:
chore(pre-commit): :pencil2: automatic fixes

signekb committed:
Merge branch 'refactor/update-extract-resource-properties-to-take-df' of https://github.com/seedcase-project/seedcase-sprout into refactor/update-extract-resource-properties-to-take-df

signekb commented on src/seedcase_sprout/core/examples.py:
```diff
@@ -214,8 +215,9 @@ def example_data_all_types() -> pl.DataFrame:
-            "my_any": pl.Series([[1]] * 3, dtype=pl.List),
-            "my_none": pl.Series([[1]] * 3, dtype=pl.List),
+            "my_list": pl.Series([[1]] * 3, dtype=pl.List),
+            "my_any": pl.Series([[1]] * 3, dtype=pl.Unknown),
```
Just thought that a “my_any” column shouldn’t be a list? So, maybe an “any” could be an Unknown?

signekb replied:
Ah, I think with your focus being on creating an example with all Frictionless types, and my focus being on creating all Polars types, this example has now become a bit of a mix. Would it be too overwhelming to create an example with all Frictionless data types and one with all Polars data types? 
Alternatively, I can create that in the tests for `extract_` instead of using this example?

signekb replied:
I’ve kept `my_any` as `pl.Unknown`, but reverted `my_none` to `pl.List` as it was before.

signekb commented on src/seedcase_sprout/core/extract_resource_properties.py:
```diff
@@ -1,46 +1,66 @@
+        FieldProperties(name=key, type=POLARS_TO_FRICTIONLESS[value.base_type()])
+        for key, value in data.schema.items()
+    ]
+    if field_properties == []:
+        raise ValueError("Failed to extract field properties from the provided data.")
```
Not sure if we want a specific error for if a df is empty / no field properties can be extracted. Thoughts?

lwjohnst86 replied:
I'm fine with not making more error types :stuck_out_tongue:

signekb committed:
docs: :pencil2: update docstring

signekb commented on tests/core/test_extract_resource_properties.py:
Have I missed tests on something? 🤔

signekb committed:
test: :white_check_mark: add test of when `DataFrame` contains missing values

signekb committed:
Merge branch 'main' into refactor/update-extract-resource-properties-to-take-df

lwjohnst86 commented on src/seedcase_sprout/core/extract_resource_properties.py:
```diff
@@ -1,46 +1,66 @@
 
-from frictionless import describe
-from frictionless.resources import JsonResource
+import seedcase_sprout.core.check_datapackage as cdp
```
```suggestion
from seedcase_sprout.core.sprout_checks.check_properties import check_resource_properties
```

We don't call `check_datapackage` outside of the Sprout specific `check_properties`

signekb replied:
Using Sprout specific checks here will fail bc the properties are incomplete.

```
  +-+---------------- 1 ----------------
    | seedcase_sprout.core.check_datapackage.check_error.CheckError: Error at `$.description` caused by `required`: 'description' is a required property
    +---------------- 2 ----------------
    | seedcase_sprout.core.check_datapackage.check_error.CheckError: Error at `$.name` caused by `required`: 'name' is a required property
    +---------------- 3 ----------------
    | seedcase_sprout.core.check_datapackage.check_error.CheckError: Error at `$.path` caused by `required`: 'path' is a required property
    +---------------- 4 ----------------
    | seedcase_sprout.core.check_datapackage.check_error.CheckError: Error at `$.title` caused by `required`: 'title' is a required property
```

signekb replied:
And as far as I understand, there’s not currently a way to ignore these using the Sprout checks. Correct me, if I'm wrong, @martonvago :)

signekb replied:
I’ve removed the check 🔥

lwjohnst86 commented on src/seedcase_sprout/core/extract_resource_properties.py:
```diff
@@ -1,46 +1,66 @@
+    resource_properties.schema.fields = _extract_field_properties(data)
 
-    properties = describe(data_path).to_dict()
+    cdp.check_resource_properties(
```
```suggestion
    check_resource_properties(
```

lwjohnst86 commented on src/seedcase_sprout/core/extract_resource_properties.py:
```diff
@@ -1,46 +1,66 @@
+    """Extract field properties from a Polars DataFrame."""
+    # Simplify by extracting the field type from the Polars base type
+    # (so e.g., Datetime(time_unit='us', time_zone='UTC') becomes Datetime)
+    field_properties = [
+        FieldProperties(name=key, type=POLARS_TO_FRICTIONLESS[value.base_type()])
+        for key, value in data.schema.items()
+    ]
```
No, the base input would need to be redesigned to be more functional. In general, with functional programming, it's the structure of the input that actually allows functional programming to work. The current `data.shema.items()` isn't designed to allow effective use of functional programming.

Plus, in general, if you need to use lambda's (outside of rare cases) in order to use functional programming like map, it suggests the overall design of the code isn't functional.

lwjohnst86 commented on src/seedcase_sprout/core/map_data_types.py:
```diff
@@ -22,3 +22,45 @@
+# Note that this mapping does not cover all Frictionless data types, but is a simplified
+# version for extracting field properties from pl.DataFrames.
+# Missing Frictionless data types are: `year`, `yearmonth`, and `geojson`.
+POLARS_TO_FRICTIONLESS = {
```
I'm not familiar enough with this area to comment :relieved:

signekb replied:
Very good point, @martonvago !

lwjohnst86 reviewed (changes requested):
Very nice! Some comments :relaxed:

lwjohnst86 commented on src/seedcase_sprout/core/extract_resource_properties.py:
```diff
@@ -1,46 +1,66 @@
 
-from seedcase_sprout.core.check_datapackage import check_resource_properties
-from seedcase_sprout.core.check_is_file import check_is_file
-from seedcase_sprout.core.check_is_supported_format import check_is_supported_format
```
Don't forget to delete this file and its tests

signekb replied:
🔥

martonvago commented:
I'll look at this this morning :sun_with_face:

martonvago reviewed (changes requested):
This all makes sense in general!
I put many comments, mostly about the type mapping, which is always very difficult. I will also have a think about how it could be nicer.

signekb commented on src/seedcase_sprout/core/extract_resource_properties.py:
```diff
@@ -1,46 +1,66 @@
+    # Simplify by extracting the field type from the Polars base type
+    # (so e.g., Datetime(time_unit='us', time_zone='UTC') becomes Datetime)
+    field_properties = [
+        FieldProperties(name=key, type=POLARS_TO_FRICTIONLESS[value.base_type()])
```
Yes, I added all Polars types from [this page](https://docs.pola.rs/api/python/stable/reference/datatypes.html) :seedling:

signekb committed:
refactor: :fire: remove `check_is_supported_format()` and its tests

signekb commented on src/seedcase_sprout/core/extract_resource_properties.py:
````diff
@@ -1,46 +1,66 @@
+        ```{python}
+        import seedcase_sprout.core as sp
+
+        extract_resource_properties(
````
Oh, yes! I wonder why I didn’t get an error about that 🤔 
Edit: Oh, this function was never added to the reference docs, that’s why!

signekb commented on src/seedcase_sprout/core/extract_resource_properties.py:
```diff
@@ -1,46 +1,66 @@
-    check_is_supported_format(data_path)
+    resource_properties = ResourceProperties()
+    resource_properties.type = "table"
+    resource_properties.schema = TableSchemaProperties()
```
Ah, good point! Thanks :sunflower:

signekb commented on tests/core/test_extract_resource_properties.py:
```diff
@@ -1,358 +1,107 @@
+        extract_resource_properties(pl.DataFrame([]))
+
+
+def test_extract_properties_with_missing_values():
+    """Test extraction when the data contains missing values."""
+    data = pl.DataFrame(
+        {
+            "col1": [1, 2, None],
+            "col2": ["a", None, "c"],
+        }
+    )
+    extracted_resource_properties = extract_resource_properties(data)
+
+    assert isinstance(extracted_resource_properties, ResourceProperties)
+    assert len(extracted_resource_properties.schema.fields) == 2
+    assert extracted_resource_properties.schema.fields[0].type == "integer"
+    assert extracted_resource_properties.schema.fields[1].type == "string"
```
```suggestion
```

signekb committed:
refactor: :recycle: apply suggestions from code review

Co-authored-by: martonvago <57952344+martonvago@users.noreply.github.com>

pre-commit-ci[bot] committed:
chore(pre-commit): :pencil2: automatic fixes

signekb committed:
Merge branch 'refactor/update-extract-resource-properties-to-take-df' of https://github.com/seedcase-project/seedcase-sprout into refactor/update-extract-resource-properties-to-take-df

signekb committed:
test: :white_check_mark: add `fields_match="equal"` to extractable properties

signekb committed:
refactor: :recycle: move empty data error to `extract_resource_properties()`

signekb committed:
docs: :memo: add `extract_resource_properties()` to reference docs

signekb committed:
Merge branch 'main' into refactor/update-extract-resource-properties-to-take-df

signekb commented on src/seedcase_sprout/core/examples.py:
```diff
@@ -214,8 +215,9 @@ def example_data_all_types() -> pl.DataFrame:
-            "my_none": pl.Series([[1]] * 3, dtype=pl.List),
+            "my_list": pl.Series([[1]] * 3, dtype=pl.List),
+            "my_any": pl.Series([[1]] * 3, dtype=pl.Unknown),
+            "my_none": pl.Series([None] * 3, dtype=pl.Null),
```
Oh, meaning “no type”, not “None”. I see!

signekb commented on src/seedcase_sprout/core/map_data_types.py:
```diff
@@ -22,3 +22,45 @@
+    pl.Datetime: "datetime",
+    pl.Duration: "duration",
+    pl.Time: "time",
+    # Nested
```
Ahh, are you suggesting that we map these nested types as Frictionless "string" then?

signekb commented on src/seedcase_sprout/core/map_data_types.py:
```diff
@@ -22,3 +22,45 @@
+    # String
+    **{type: "string" for type in [pl.String, pl.Categorical, pl.Enum, pl.Utf8]},
+    # Other
+    pl.Binary: "binary",
```
Ah, you're right! It could be strings, but with format="binary", according to https://datapackage.org/standard/table-schema/#string

signekb commented on src/seedcase_sprout/core/map_data_types.py:
```diff
@@ -22,3 +22,45 @@
+    # Temporal
+    pl.Date: "date",
+    pl.Datetime: "datetime",
+    pl.Duration: "duration",
```
Hmmm… You’re right. It seems that [pl.Duration](https://docs.pola.rs/api/python/stable/reference/api/polars.datatypes.Duration.html#polars.datatypes.Duration) is actually an integer. So, maybe mapping to “string” is fine for now?

signekb commented on src/seedcase_sprout/core/map_data_types.py:
```diff
@@ -22,3 +22,45 @@
+    pl.Time: "time",
+    # Nested
+    pl.Array: "array",
+    pl.List: "any",
```
I actually think, it could just be list! It seems that both Polars and Frictionless has a "list" data type that cannot be nested, like `[1, 2, [3, 4]]`

martonvago replied:
I think Frictionless `list` is v2 and hasn't yet been included in the JSON schema. Or last time I checked it wasn't there. So that means the checks won't recognise it and will flag it as an unknown type.

signekb commented on tests/core/test_extract_resource_properties.py:
I'll create a separate dataframe + resource properties for this. I was trying to use `examples_data_all_type()` for this, but didn't know that was for all *Frictionless* types :+1:

signekb committed:
refactor: :recycle: apply suggestions from code review

signekb commented on tests/core/test_extract_resource_properties.py:
```diff
@@ -1,358 +1,107 @@
-        (empty_data, schema_empty_data),
-        (tidy_data, schema_tidy_data),
-        (non_tidy_data, schema_non_tidy_data),
+        (example_data(), _keep_extractable_properties(example_resource_properties())),
```
I’ll change this to use data + resources with all Polars data types :+1:

signekb committed:
docs: :pencil2: add "Frictionless" to docstring

signekb committed:
revert: :rewind: change `my_any` back to `pl.List`

Since adding `dtype=pl.Unknown` didn't stop Polars from changing it to `List` in the schema.

signekb committed:
feat: :sparkles: add example data and resource props with all Polars data types

signekb committed:
test: :white_check_mark: use example data and resource properties with all Polars data types

Instead of the example with all Frictionless data types

signekb committed:
docs: :pencil2: add `geopoint` to list of missing Frictionless data types

signekb committed:
docs: :pencil2: Move comment to where it could be implemented

signekb committed:
Merge branch 'main' into refactor/update-extract-resource-properties-to-take-df

pre-commit-ci[bot] committed:
chore(pre-commit): :pencil2: automatic fixes

signekb commented:
This is as far as I got before leaving for Easter break 🐣 There are still some unresolved comments.
Happy/Merry Easter!

martonvago commented:
I did some work on the maps in #1284 , incorporating everyone's thoughts / requirements :world_map:

lwjohnst86 commented on src/seedcase_sprout/core/extract_resource_properties.py:
```diff
@@ -1,46 +1,67 @@
+    resource_properties.schema.fields = _extract_field_properties(data)
 
-    properties.pop("dialect", None)
+    cdp.check_resource_properties(
+        resource_properties,
+    )
```
```suggestion
```

See above. As @signekb said, we know they aren't complete yet, so we shouldn't check at this point.

lwjohnst86 commented on src/seedcase_sprout/core/map_data_types.py:
```diff
@@ -22,3 +22,44 @@
+# Note that this mapping does not cover all Frictionless data types, but is a simplified
+# version for extracting field properties from `pl.DataFrame`s.
+# Missing Frictionless data types are: `year`, `yearmonth`, `geopoint, and `geojson`.
+POLARS_TO_FRICTIONLESS: dict[pl.DataType, FieldType] = {
```
When #1284 is done, I will finish up this PR to get it done and merged.

signekb replied:
@lwjohnst86 I can do that? :)

lwjohnst86 replied:
I was optimistic that I could get to it a day after the comment... It's been 5 now, so clearly very optimistic 😂😂

signekb committed:
Merge branch 'main' into refactor/update-extract-resource-properties-to-take-df

signekb committed:
fix: :bug: update uv lock after merge

signekb committed:
refactor: :recycle: update example properties to fit polars to dp mapping

signekb committed:
refactor: :recycle: remove check of extracted resources

We know that they will be incomplete--and therefore fail the check--at this point.

signekb committed:
refactor: :recycle: use new mapping and map functions to extract properties

signekb committed:
test: :white_check_mark: restructure test to follow given, when, then

signekb committed:
docs: :memo: Frictionless -> DataPackage in docstring

signekb committed:
fix: :truck: rename to `_polars_to_datapackage()` instead of `_frictionless()`

signekb commented on src/seedcase_sprout/core/map_data_types.py:
I only merged with `main` and changed `_polars_to_frictionless()` to `_polars_to_datapackage()` in this file.

signekb committed:
docs: :pencil2: "Frictionless" -> "Data Package"

signekb committed:
Merge branch 'main' into refactor/update-extract-resource-properties-to-take-df

lwjohnst86 commented on src/seedcase_sprout/core/examples.py:
```diff
@@ -226,13 +226,110 @@ def example_data_all_types() -> pl.DataFrame:
+                FieldProperties(name="my_time", type="time"),
+                FieldProperties(name="my_array", type="array"),
+                # We can't map to DataPackage `list` bc it's missing in the TableSchema
+                # profile we use generate the datapackage schema we check properties
```
```suggestion
                # profile we use to generate the datapackage schema that we check properties
```

lwjohnst86 reviewed (changes requested):
Nice! only small changes that I will do and then approve and merge :grin:

lwjohnst86 committed:
refactor: :recycle: suggestions from review

lwjohnst86 committed:
Merge branch 'main' of https://github.com/seedcase-project/seedcase-sprout into refactor/update-extract-resource-properties-to-take-df

lwjohnst86 committed:
style: :art: line too long, shortened

lwjohnst86 committed:
docs: :memo: need to add the resource path

lwjohnst86 committed:
refactor: :recycle: revise example resource name, including with correct path

lwjohnst86 committed:
revert: :rewind: this does actually need a lambda sadly
